logstash-core/src/main/java/org/logstash/common/io/DeadLetterQueueWriter.java

```
@@ -172,42 +188,121 @@ private static boolean alreadyProcessed(final Event event) {
         } catch (IOException e){
-            logger.debug("Unable to delete lock file", e);
+            logger.debug("Unable to delete fileLock file", e);
         }
     }
 
-    public boolean isOpen() {
-        return open.get();
+    private void nextWriter() throws IOException {
+        currentWriter = new RecordIOWriter(queuePath.resolve(String.format(TEMP_FILE_PATTERN, ++currentSegmentIndex)));
+        currentQueueSize.increment();
     }
 
-    public Path getPath(){
-        return queuePath;
+    // Clean up existing temp files - files with an extension of .log.tmp. Either delete them if an existing
+    // segment file with the same base name exists, or rename the
+    // temp file to the segment file, which can happen when a process ends abnormally
+    private void cleanupTempFiles() throws IOException {
+        DeadLetterQueueWriter.listFiles(queuePath, ".log.tmp")
+                .forEach(this::cleanupTempFile);
     }
 
-    public long getCurrentQueueSize() {
-        return currentQueueSize.longValue();
+    private static Stream<Path> listFiles(Path path, String suffix) throws IOException {
+        try(final Stream<Path> files = Files.list(path)) {
+            return files.filter(p -> p.toString().endsWith(suffix))
+                    .collect(Collectors.toList()).stream();
+        }
+    }
+
+    // check if there is a corresponding .log file - if yes delete the temp file, if no atomic move the
+    // temp file to be a new segment file..
+    private void cleanupTempFile(final Path file) {
+        String filename = file.getFileName().toString().split("\\.")[0];
+        Path segmentFile = queuePath.resolve(String.format("%s.log", filename));
+        try {
+            if (Files.exists(segmentFile)) {
+                Files.delete(file);
+            }
+            else {
+                if (RecordIOReader.validNonEmptySegment(file)) {
+                    Files.move(file, segmentFile, StandardCopyOption.ATOMIC_MOVE);
```

what if the segment file in invalid? should we log it?

---

We do - https://github.com/elastic/logstash/blob/daef85e9b80edf16a585eebf66ec3d5792f3081e/logstash-core/src/main/java/org/logstash/common/io/RecordIOReader.java#L369 although I wonder if it should probably be an `error` rather than a `warn`. WDYT?

---

yeah ... it is an error but we do ignore it in the logic, not sure.

---

ok so my question here is what do we do when there is an *invalid* temp file around? currently only a WARN log `Error reading segment file ...` will be outputted but should we also give more information on the problem? I think we should explain to the user something like: we found a temps segment file but it does not look valid, we will ignore the error and leave the file in place. 

Can this create a problem for the DLQ reader if there is a break in the file sequence?

---

I like the idea of keeping the file around, maybe even renaming to `err` or similar, to help with forensics if an issue arises.

There shouldn't be an issue if there is a break in the file sequence, the sequence numbers aren't used for ordering

---

agree about leaving the file around, my comment was more about giving more context and information in the logs for the user to understand what is happening?

---

Yes, I will do that too